From ea9e71000e0d1ebcdd73771e39fd567b25e04f54 Mon Sep 17 00:00:00 2001
From: Andy CrossGate Yan <GeForce8800Ultra@gmail.com>
Date: Tue, 11 Aug 2026 09:23:22 +0000
Subject: [PATCH] UI: Revive navbar layout tuning via sysui_nav_bar tunable

[Port to A16] LineageOS 23 dropped the shared NAV_BAR_INVERSE/navigation-hint
ContentObserver this patch hooked into, so the observer is recreated - but with
the A14 lifecycle: field observer built in the constructor, registered in
onAttachedToWindow and unregistered in onDetachedFromWindow (the A16 class
already has onDetachedFromWindow), instead of an unregistered constructor-time
registration that would leak one observer per navbar inflation.

Change-Id: I3ac611b04ea00a675cea836e840ffe0fb241c11a
---
 .../views/NavigationBarInflaterView.java      | 38 +++++++++++++++++++
 1 file changed, 38 insertions(+)

diff --git a/packages/SystemUI/src/com/android/systemui/navigationbar/views/NavigationBarInflaterView.java b/packages/SystemUI/src/com/android/systemui/navigationbar/views/NavigationBarInflaterView.java
index fe2f3890bfa8..a8899a594e9f 100644
--- a/packages/SystemUI/src/com/android/systemui/navigationbar/views/NavigationBarInflaterView.java
+++ b/packages/SystemUI/src/com/android/systemui/navigationbar/views/NavigationBarInflaterView.java
@@ -22,6 +22,9 @@ import static android.view.WindowManagerPolicyConstants.NAV_BAR_MODE_3BUTTON;
 import android.annotation.Nullable;
 import android.content.Context;
 import android.content.res.Configuration;
+import android.database.ContentObserver;
+import android.net.Uri;
+import android.provider.Settings;
 import android.graphics.drawable.Icon;
 import android.util.AttributeSet;
 import android.util.Log;
@@ -116,6 +119,9 @@ public class NavigationBarInflaterView extends FrameLayout {
 
     private boolean mIsVertical;
     private boolean mAlternativeOrder;
+    private boolean mUsingCustomLayout;
+
+    private final ContentObserver mContentObserver;
 
     private LauncherProxyService mLauncherProxyService;
     private int mNavBarMode = NAV_BAR_MODE_3BUTTON;
@@ -126,6 +132,15 @@ public class NavigationBarInflaterView extends FrameLayout {
         mLauncherProxyService = Dependency.get(LauncherProxyService.class);
         mListener = new Listener(this);
         mNavBarMode = Dependency.get(NavigationModeController.class).addListener(mListener);
+        mContentObserver = new ContentObserver(null) {
+            @Override
+            public void onChange(boolean selfChange, @Nullable Uri uri) {
+                if (Settings.Secure.getUriFor(NAV_BAR_VIEWS).equals(uri)) {
+                    setNavigationBarLayout(Settings.Secure.getString(mContext.getContentResolver(),
+                            NAV_BAR_VIEWS));
+                }
+            }
+        };
     }
 
     @VisibleForTesting
@@ -169,13 +184,26 @@ public class NavigationBarInflaterView extends FrameLayout {
         mNavBarMode = mode;
     }
 
+    @Override
+    protected void onAttachedToWindow() {
+        super.onAttachedToWindow();
+        Uri navBarViews = Settings.Secure.getUriFor(NAV_BAR_VIEWS);
+        mContext.getContentResolver().registerContentObserver(navBarViews, false,
+                mContentObserver);
+        mContentObserver.onChange(true, navBarViews);
+    }
+
     @Override
     protected void onDetachedFromWindow() {
         Dependency.get(NavigationModeController.class).removeListener(mListener);
+        mContext.getContentResolver().unregisterContentObserver(mContentObserver);
         super.onDetachedFromWindow();
     }
 
     public void onLikelyDefaultLayoutChange() {
+        // Don't override custom layouts
+        if (mUsingCustomLayout) return;
+
         // Reevaluate new layout
         final String newValue = getDefaultLayout();
         if (!Objects.equals(mCurrentLayout, newValue)) {
@@ -184,6 +212,16 @@ public class NavigationBarInflaterView extends FrameLayout {
         }
     }
 
+    public void setNavigationBarLayout(String layoutValue) {
+        mContext.getMainExecutor().execute(() -> {
+            if (!Objects.equals(mCurrentLayout, layoutValue)) {
+                mUsingCustomLayout = layoutValue != null;
+                clearViews();
+                inflateLayout(layoutValue);
+            }
+        });
+    }
+
     public void setButtonDispatchers(SparseArray<ButtonDispatcher> buttonDispatchers) {
         mButtonDispatchers = buttonDispatchers;
         clearDispatcherViews();
-- 
2.43.0

